ykbh/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+#![feature(test)]
+extern crate test;
+
+use ykpack::{self, Local, Terminator, Statement, IPlace, Constant, ConstantInt, UnsignedInt};
+use yktrace::sir::SIR;
+use std::convert::TryInto;
+
+const ZST: Object = Object { data: Vec::new() };
+
+#[derive(Clone, Debug)]
+/// A data object, storing bytes.
+pub struct Object {
+    data: Vec<u8>
+}
+
+impl Object {
+
+    /// Creates a new Object from a `usize`.
+    fn from_usize(ptr: usize) -> Object {
+        let data = ptr.to_be_bytes();
+        Object {
+            data: Vec::from(data)
+        }
+    }
+
+    /// Creates a new Object from a `u8`.
+    fn from_u8(val: u8) -> Object {
+        Object {
+            data: Vec::from([val])
+        }
+    }
+
+    /// Converts the objects data back to a `usize`.
+    fn to_usize(&self) -> usize {
+        usize::from_be_bytes(self.data.as_slice().try_into().expect("Unexpected data length."))
+    }
+
+    /// Given a pointer and size, creates a new object by copying `size` data from `ptr`.
+    fn from_ptr(ptr: usize, size: usize) -> Object {
+        let mut data = Vec::with_capacity(size);
+        unsafe {
+            std::ptr::copy(ptr as *const u8, data.as_mut_ptr(), size);
+            data.set_len(size);
+        }
+        Object {
+            data
+        }
+    }
+}
+
+pub struct SIRInterpreter {
+    vars: Vec<Option<Object>>,
+}
+
+impl SIRInterpreter {
+    pub fn new(vars: Vec<Option<Object>>) -> Self {
+        SIRInterpreter {
+            vars,
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
```

Comment:
ANDing bits to detect something feels quite 1980s: shouldn't `ykpack` be exposing some sort of nice function for this?

Reply:
There is the `bitflags` crate.

Reply:
I recall looking at the bitflags crate when introducing these flags. I can't remember why I chose not to use it.

Reply:
That's not really what I mean. What I mean is: how we represent flags internally shouldn't be exposed externally. There should almost certainly be a *method* called something like `trace_debug()` so that we can change the internal representation later. Sound like a plan?

Reply:
We can do that, but I doubt we will change the representation.

Will raise an issue anyway.

Reply:
No-one thinks they will change a representation... until they do ;)